### PR TITLE
Added some tests for version restore functionality

### PR DESCRIPTION
**Pull request by danrot**

### CODE REVIEW

File: tests/15_Versioning/VersionManagerTest.php [hunk 1]
```diff
@@ -275,6 +283,39 @@ public function testRestorePendingChanges()
...
+        $this->vm->checkout($nodePath);
+        $this->assertTrue($this->vm->isCheckedOut($nodePath));
+
+        $this->vm->restore(true, $version);
+        $this->assertFalse($this->vm->isCheckedOut($nodePath));
+    }
+
+    public function testRestoreBaseProperties()
+    {
+        // TODO also check for primary node type and uuid once it can be changed
```
Comment (danrot) [inline tests/15_Versioning/VersionManagerTest.php]: @dbu I am not sure about this one. The specification says how to handle these properties on restore, but if I try to write a test for it jackalope says it's not allowed to change these properties. Am I missing something? Is this just a rarely happening edge case?

Comment (dbu) [inline tests/15_Versioning/VersionManagerTest.php]: i would think that the primary type never changes, or can it? if it does not change it also does not need to be reverted...

and the uuid of the original node should stay the same when restoring, no?

File: tests/15_Versioning/VersionManagerTest.php [hunk 2]
```diff
@@ -275,6 +283,39 @@ public function testRestorePendingChanges()
...
+
+        $this->session->save();
+
+        $node = $this->session->getNode($nodePath);
+        $this->assertContains('mix:created', $node->getPropertyValue('jcr:mixinTypes'));
+
+        $this->vm->restore(true, $version);
+
+        $node = $this->session->getNode($nodePath);
+        $this->assertEquals(array('mix:versionable'), $node->getPropertyValue('jcr:mixinTypes'));
```
Comment (dbu) [inline tests/15_Versioning/VersionManagerTest.php]: you could also assert that there is no `created` property anymore

**Comment by dbu** [comment]
jackalope-jackrabbit is happy with these tests so far. so the assumptions seem correct.

File: tests/15_Versioning/VersionManagerTest.php (hunk 1 above)
Comment (danrot) [inline tests/15_Versioning/VersionManagerTest.php]: I've got the same understanding until now, but after having a look at the specification I am not sure if there are some edge cases at which this properties can change. Otherwise 15.7.2 on http://www.day.com/specs/jcr/2.0/15_Versioning.html doesn't make any sense to me.

File: tests/15_Versioning/VersionManagerTest.php (hunk 2 above)
Comment (danrot) [inline tests/15_Versioning/VersionManagerTest.php]: But then it might be possible that all mixinTypes are gone, which would be a wrong behavior. Maybe I also should add the complete array check instead of `assertContains` a few lines earlier?

File: tests/15_Versioning/VersionManagerTest.php (hunk 1 above)
Comment (dbu) [inline tests/15_Versioning/VersionManagerTest.php]: i would propose we assume the uuid immutable and you can add a check
that throws an exception when the frozen uuid is not the one of the
current node, saying something unexpected happened...

for mixins, they certainly can change.

about the primary type, this seems to be another jackrabbit restriction
that is actually a mistake https://github.com/jackalope/jackalope/issues/247

but in my opinion we can just leave a TODO here referencing that issue,
and a comment in that issue that it would also have to be fixed in
version restore if this is ever changed. as its now, its not possible to
change the primary type so no need (yet) to restore it.

File: tests/15_Versioning/VersionManagerTest.php (hunk 2 above)
Comment (dbu) [inline tests/15_Versioning/VersionManagerTest.php]: i mean the jcr:created property that the mix:created made jackalope write.

for the mixins i think its fine like this, or you will need to make sure not to rely on a particular order of the mixins as that is not determined...

**Comment by danrot** [comment]
Yeah, I write them in jackalope-jackrabbit first, and take over the changes and required implementations to jackalope-doctrine-dbal afterwards :-)

File: tests/15_Versioning/VersionManagerTest.php (hunk 2 above)
Comment (danrot) [inline tests/15_Versioning/VersionManagerTest.php]: Is it possible that the properties are not deleted after the mixin is removed? Because it seems to me like that... And actually that would also make sense to me.

File: tests/15_Versioning/VersionManagerTest.php [hunk 3]
```diff
@@ -275,6 +282,113 @@ public function testRestorePendingChanges()
...
+
+        $this->assertCount(1, $node->getNodes());
+        $this->assertEquals('child1', $node->getNode('childNode2')->getPropertyValue('foo'));
+
+        $this->vm->restore(false, $version);
+        $node = $this->session->getNode($nodePath);
+
+        $this->assertCount(2, $node->getNodes());
+        $this->assertEquals('child1', $node->getNode('childNode1')->getPropertyValue('foo'));
+        $this->assertEquals('child2', $node->getNode('childNode2')->getPropertyValue('foo'));
```
Comment (danrot) [inline tests/15_Versioning/VersionManagerTest.php]: @dbu @lsmith77 This is failing for jackrabbit. But I guess this should be the behavior as I read in 15.7.6 in http://www.day.com/specs/jcr/2.0/15_Versioning.html. It basically says that each child from the frozen node should replace any existing child in the actual node.

Comment (danrot) [inline tests/15_Versioning/VersionManagerTest.php]: This is interesting, it looks to me like this is an error in the jackalope jackrabbit transport layer, since jackrabbit itself (I have tested it via the cli in jackrabbit-standalone.jar) behaves as expected, as you can see in the following script:

```
[/] > addnode test

elapsed time: 25 ms.

[/*] > addmix

addmix     addmixin
[/*] > addmixin mix:versionable test

elapsed time: 2 ms.

[/*] > cd test

elapsed time: 0 ms.

[/test*] > addnode childnode1

elapsed time: 1 ms.

[/test*] > cd childnode1

elapsed time: 0 ms.

[/test/childnode1*] > setprop

setprop               setproperty           setpropertyfromfile   setpropfromfile
[/test/childnode1*] > setprop foo child1

elapsed time: 2 ms.

[/test/childnode1*] > cd ..

elapsed time: 0 ms.

[/test*] > addnode childnode2

elapsed time: 1 ms.

[/test*] > cd childnode

childnode1   childnode2
[/test*] > cd childnode2

elapsed time: 0 ms.

[/test/childnode2*] > setprop

setprop               setproperty           setpropertyfromfile   setpropfromfile
[/test/childnode2*] > setprop foo child2

elapsed time: 0 ms.

[/test/childnode2*] > cd ..

elapsed time: 0 ms.

[/test*] > save

elapsed time: 74 ms.

[/test] > checkin

elapsed time: 118 ms.

[/test] > cd childnode

childnode1   childnode2
[/test] > cd childnode1

elapsed time: 0 ms.

[/test/childnode1] > lsprop
name                           multiple  type      length   preview
------------------------------ --------- --------- -------- ------------------
jcr:primaryType                false     Name      15       nt:unstructured
foo                            false     String    6        child1


total
2 nodes

elapsed time: 3 ms.

[/test/childnode1] > cd ../childnode2

elapsed time: 0 ms.

[/test/childnode2] > lsprop
name                           multiple  type      length   preview
------------------------------ --------- --------- -------- ------------------
foo                            false     String    6        child2
jcr:primaryType                false     Name      15       nt:unstructured


total
2 nodes

elapsed time: 1 ms.

[/test/childnode2] > cd ..

elapsed time: 0 ms.

[/test] > checkout

elapsed time: 19 ms.

[/test] > rm childnode

childnode1   childnode2
[/test] > rm childnode1

elapsed time: 2 ms.

[/test*] > cd childnode2

elapsed time: 0 ms.

[/test/childnode2*] > setprop

setprop               setproperty           setpropertyfromfile   setpropfromfile
[/test/childnode2*] > setprop foo child1

elapsed time: 1 ms.

[/test/childnode2*] > save

elapsed time: 18 ms.

[/test/childnode2] > cd ..

elapsed time: 0 ms.

[/test] > ls
name                           type            node      new       modified
------------------------------ --------------- --------- --------- ---------
childnode2                     nt:unstructured true      false     false
jcr:isCheckedOut               Boolean         false     false     false
jcr:versionHistory             Reference       false     false     false
jcr:mixinTypes                 Name            false     false     false
jcr:predecessors               Reference       false     false     false
jcr:baseVersion                Reference       false     false     false
jcr:uuid                       String          false     false     false
jcr:primaryType                Name            false     false     false

total

elapsed time: 5 ms.

[/test] > cd childnode2

elapsed time: 1 ms.

[/test/childnode2] > lsprop
name                           multiple  type      length   preview
------------------------------ --------- --------- -------- ------------------
foo                            false     String    6        child1
jcr:primaryType                false     Name      15       nt:unstructured


total
2 nodes

elapsed time: 1 ms.

[/test/childnode2] > cd ..

elapsed time: 0 ms.

[/test] > lsversions
version              labels
-------------------- --------------------------------------------------
jcr:rootVersion
1.0

elapsed time: 32 ms.

[/test] > restore 1.0

elapsed time: 32 ms.

[/test] > ls
name                           type            node      new       modified
------------------------------ --------------- --------- --------- ---------
childnode1                     nt:unstructured true      false     false
childnode2                     nt:unstructured true      false     false
jcr:isCheckedOut               Boolean         false     false     false
jcr:versionHistory             Reference       false     false     false
jcr:mixinTypes                 Name            false     false     false
jcr:predecessors               Reference       false     false     false
jcr:baseVersion                Reference       false     false     false
jcr:uuid                       String          false     false     false
jcr:primaryType                Name            false     false     false

total

elapsed time: 13 ms.

[/test] > cd childnode

childnode1   childnode2
[/test] > cd childnode1

elapsed time: 0 ms.

[/test/childnode1] > lsprop
name                           multiple  type      length   preview
------------------------------ --------- --------- -------- ------------------
jcr:primaryType                false     Name      15       nt:unstructured
foo                            false     String    6        child1
jcr:mixinTypes                 true[0]   Name      0


total
3 nodes

elapsed time: 1 ms.

[/test/childnode1] > cd ../childnode2

elapsed time: 1 ms.

[/test/childnode2] > lsprop
name                           multiple  type      length   preview
------------------------------ --------- --------- -------- ------------------
jcr:mixinTypes                 true[0]   Name      0
foo                            false     String    6        child2
jcr:primaryType                false     Name      15       nt:unstructured


total
3 nodes

elapsed time: 2 ms.
```

File: tests/15_Versioning/VersionManagerTest.php (hunk 2 above)
Comment (dbu) [inline tests/15_Versioning/VersionManagerTest.php]: i would expect that if the frozen node has no jcr:created property and it is restored, that property should be removed.

are you saying jackrabbit does not do that? even when refreshing the session after restoring?

Comment (danrot) [inline tests/15_Versioning/VersionManagerTest.php]: I thought so, it has been some time since I've checked the last time :wink: 

Do you really would expect that? Isn't it possible that the `jcr:created` property has been there before the mixin has been added? Would feel strange to me if removing the mixin is then also removing the property...

File: tests/15_Versioning/VersionManagerTest.php (hunk 3 above)
Comment (danrot) [inline tests/15_Versioning/VersionManagerTest.php]: @dbu I am missing a comment of yours on this one :-)

File: tests/15_Versioning/VersionManagerTest.php (hunk 2 above)
Comment (dbu) [inline tests/15_Versioning/VersionManagerTest.php]: is jcr:created not backed up? if it is backed up, restoring the old version should have deleted it like it deletes other properties that where not existing when the version was created.

File: tests/15_Versioning/VersionManagerTest.php (hunk 3 above)
Comment (dbu) [inline tests/15_Versioning/VersionManagerTest.php]: i will have to investigate this.

Comment (dbu) [inline tests/15_Versioning/VersionManagerTest.php]: ok. the expectations seem correct, this is a problem with jackalope-jackrabbit not noticing that the children also changed. fixed in https://github.com/jackalope/jackalope/pull/294

can you check if things still work out when you have these changes with your code? i wonder if marking nodes as dirty could cause side effects when you restore nodes in memory instead.

File: tests/15_Versioning/VersionManagerTest.php (hunk 2 above)
Comment (dbu) [inline tests/15_Versioning/VersionManagerTest.php]: can you please check this bit? otherwise i think this would be ready to merge, the issue below is indeed a bug in the code and the test is correct.

Comment (danrot) [inline tests/15_Versioning/VersionManagerTest.php]: Have added this check.

**Comment by dbu** [comment]
rebased on master and will create a new PR

**Comment by dbu** [comment]
see #166
